### PR TITLE
ENH: Use pytest to find and run tests

### PR DESCRIPTION
**Pull request by aragilar**
Part of a plan to move from unittest to pytest (giving nicer reporting, use of pdb for debugging, and hopefully easier testing). ~~WIP until I check there are no regressions, and that the setup is robust against pytest changes.~~ Should now be fine to use.

### CODE REVIEW

**Comment by aragilar** [comment]
Comments on the setup used:
 * Running pytest in the existing python process (as earlier versions of the code did, and how the tests are currently run with unittest) causes incorrect values for coverage, and a lot of pain when trying to work on windows with additional pytest arguments (and with additional arguments generally).
 * Coverage is collected, but currently there's no online coverage service connected to CI, opinions on what's the best service to use, if any (coveralls?, codecov?)?
 * I haven't added any tests yet to ensure that `h5py.run_tests()` keeps working, I could add an additional tox which runs setup.py test (as that depends on `h5py.run_tests()` and we're not checking that either, do people thing that's sufficient?

**Comment by tacaswell** [comment]
matplotlib uses codecov and have been pretty happy with it.

File: h5py/tests/__init__.py [hunk 1]
```diff
@@ -7,35 +7,18 @@
...
-
-    return result
+from __future__ import absolute_import
+
+
+def run_tests(args=''):
+    try:
+        from pytest import main
+    except ImportError:
+        print("Tests require pytest, pytest not installed")
```
Comment (takluyver) [inline h5py/tests/__init__.py]: This should probably have a `return 1` so that the exit code indicates something went wrong.

Comment (aragilar) [reply h5py/tests/__init__.py]: Good catch, I've added that now.

File: .gitignore [hunk 2]
```diff
@@ -24,4 +24,3 @@ MANIFEST
 .eggs/
 *.so
 *~
-.pytest_cache
```
Comment (takluyver) [inline .gitignore]: This is being removed from `.gitignore`? Shouldn't it be there? I have something similar for other projects using pytest.

**Comment by takluyver** [comment]
:+1:

File: .gitignore (hunk 2 above)
Comment (aragilar) [reply .gitignore]: That was me screwing up the rebase, I've dropped that change (so `.pytest_cache` is still being ignored).

**Pull request merged**